src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,425 @@
+                {
+                    // Example: missing final `)` in `do { } while (x$$`
+                    return;
+                }
+            }
+
+            // Move to space after the last delimiter
+            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(GetEndPosition(root, lastDelimiterSpan.End, currentNode.Kind())));
+        }
+
+        /// <summary>
+        /// Examines the enclosing statement-like syntax for an expression which is eligible for statement completion.
+        /// </summary>
+        /// <remarks>
+        /// <para>This method tries to identify <paramref name="currentNode"/> as a node located within an argument
+        /// list, where the immediately-containing statement resembles an "expression statement". This method returns
+        /// <see langword="true"/> if the node matches a recognizable pattern of this form.</para>
+        /// </remarks>
+        private static bool LooksLikeNodeInArgumentListForStatementCompletion(SyntaxNode currentNode, ISyntaxFactsService syntaxFacts, SnapshotPoint? caret)
+        {
+            while (!currentNode.IsKind(SyntaxKind.ArgumentList, SyntaxKind.ArrayRankSpecifier))
+            {
+                if (currentNode == null)
+                {
+                    // We reached the root without detecting a node of interest.
+                    return false;
+                }
+
+                if (currentNode.IsKind(SyntaxKind.InterpolatedStringExpression, SyntaxKind.StringLiteralExpression))
+                {
+                    // No special action is performed at this time if `;` is typed inside a string, including
+                    // interpolated strings.  
+                    // If caret is at the end of the line, it is outside the string so is a candidate for 
+                    // statement completion
+                    if (caret != null && IsCaretAtEndOfLine((SnapshotPoint)caret))
+                    {
+                        return true;
+                    }
+
+                    return false;
+                }
+
+                if (syntaxFacts.IsStatement(currentNode)
+                    || currentNode.IsKind(SyntaxKind.VariableDeclaration))
+                {
+                    // We reached an enclosing statement-like syntax without finding an intermediate argument-list-like
+                    // syntax. Do not treat this statement as a candidate for statement completion because inserting a
+                    // semicolon at the current location is likely to create a valid statement.
+                    //
+                    // IsStatement: The syntax is a known statement syntax
+                    // VariableDelaration: The expression is part of a field initializer, which is not part of a
+                    //      statement syntax but behaves like an expression statement for the purposes of statement
+                    //      completion.
+                    return false;
```

Comment:
i cna't reconcile this comment with the method name and return value.  THe method asks if this LooksLikeNodeInArgumentListForStatementCompletion.  THe comment then says that this does "behave likes an expressoin statement for the purposes of statement completion".  But then it returns 'false'.  All in all, it'm just confused :)